File: Ahorn/entities/InGameMapSubAreaController.jl
module XaphanHelperInGameMapSubAreaController

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/InGameMapSubAreaController" InGameMapSubAreaController(x::Integer, y::Integer, exploredRoomColor::String="D83890", unexploredRoomColor::String="000080", secretRoomColor::String="057A0C", heatedRoomColor::String="FF650D", roomBorderColor::String="FFFFFF", elevatorColor::String="F80000", subAreaName::String="", subAreaIndex::Number=0)

const placements = Ahorn.PlacementDict(
    "In-Game Map Sub Area Controller (Xaphan Helper)" => Ahorn.EntityPlacement(
        InGameMapSubAreaController
    )
)

function Ahorn.selection(entity::InGameMapSubAreaController)
    x, y = Ahorn.position(entity)

    return Ahorn.Rectangle(x - 10, y - 10, 20, 20)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::InGameMapSubAreaController, room::Maple.Room)
    Ahorn.Cairo.save(ctx)

    Ahorn.set_antialias(ctx, 1)
    Ahorn.set_line_width(ctx, 1);

    Ahorn.drawCircle(ctx, 0, 0, 10, (1.0, 1.0, 1.0, 1.0))

    Ahorn.Cairo.restore(ctx)

    Ahorn.drawSprite(ctx, "util/XaphanHelper/subAreaController", 0, 0)
end

end

File: Loenn/entities/InGameMapSubAreaController.lua
local InGameMapSubAreaController = {}

InGameMapSubAreaController.name = "XaphanHelper/InGameMapSubAreaController"
InGameMapSubAreaController.depth = -100000
InGameMapSubAreaController.fieldOrder = {
    "x", "y", "subAreaIndex", "subAreaName", "exploredRoomColor", "unexploredRoomColor", "secretRoomColor", "heatedRoomColor", "roomBorderColor", "elevatorColor"
}
InGameMapSubAreaController.fieldInformation = {
    roomIndicatorColor = {
        subAreaIndex = "integer"
    },
    exploredRoomColor = {
        fieldType = "color"
    },
    unexploredRoomColor = {
        fieldType = "color"
    },
    secretRoomColor = {
        fieldType = "color"
    },
    heatedRoomColor = {
        fieldType = "color"
    },
    roomBorderColor = {
        fieldType = "color"
    },
    elevatorColor = {
        fieldType = "color"
    }
}
InGameMapSubAreaController.placements = {
    name = "InGameMapSubAreaController",
    data = {
        subAreaIndex = 0,
        subAreaName = "",
        exploredRoomColor = "D83890",
        unexploredRoomColor = "000080",
        secretRoomColor = "057A0C",
        heatedRoomColor = "FF650D",
        roomBorderColor = "FFFFFF",
        elevatorColor = "F80000",
    }
}

InGameMapSubAreaController.texture = "util/XaphanHelper/Loenn/subAreaController"

return InGameMapSubAreaController